stan/math/prim/mat/fun/trace_gen_quad_form.hpp

```diff
@@ -5,12 +5,34 @@
 #include <stan/math/prim/meta.hpp>
 #include <stan/math/prim/mat/err/check_multiplicable.hpp>
 #include <stan/math/prim/mat/err/check_square.hpp>
+#include <stan/math/prim/mat/fun/trace.hpp>
+#include <stan/math/prim/mat/fun/multiply.hpp>
+#include <stan/math/prim/mat/fun/transpose.hpp>
+#include <exception>
 
 namespace stan {
 namespace math {
 /**
- * Compute trace(D B^T A B).
- **/
+ * Return the trace of D times the quadratic form of B and A.
+ * That is, `trace_gen_quad_form(D, A, B) = trace(D * B' * A * B).`
+ *
+ * @tparam TD type of first argument scalar
+ * @tparam RD row type of first argument
```

Comment:
I'm happy to reword.  There's certainly support for @rok-cesnovar's proposal because here's the general template from Eigen in their [quickstart guide](https://eigen.tuxfamily.org/dox/group__QuickRefPage.html).

```
typedef Matrix<Scalar, RowsAtCompileTime, ColsAtCompileTime, Options> MyMatrixType;
```

But I like their [doc on the template parameters](https://eigen.tuxfamily.org/dox/classEigen_1_1Matrix.html) better:

```
Number of rows, or Dynamic
```

The only values we allow for the row template parameter are `1` for a single row fixed at compile time (and hence a different template specialization) or `-1` (aka `Eigen::Dynamic`), which means the number of rows are determined at run time.  We should change all of our doc for matrices to follow this pattern!

Reply:
Agreed. Lets tackle this Stan Math wide, so this is out of scope for this PR.

Reply:
fixed in this file;  opened issue https://github.com/stan-dev/stan/issues/2855 to fix elsewhere.

Reply:
Great, thanks!